### DISCUSSION TITLE
Multiple Data Sources

### DISCUSSION THREAD

**Discussion by MichalisDBA** [9tigerio/db2rest#458] [Ideas]
Hello.

I do not know if you have this on the roadmap. I'll just give you my idea on how i envision to work with multiple data sources running a single instance of db2rest.

Create a config file. Let's say `datasources.conf`

Inside this configuration file have the following settings:

```
db2rest:
	SERVER_PORT: 8090

java:
	-Xms2048m
	-Xmx8192m

db1:
	service:
		DRIVER: ORACLE
		USERNAME: USER
		PASSWORD: PASS
		SERVER: IP / FQDN
		INSTANCE: MY_INSTANCE
		SCHEMA: MY_SCHEMA #optional
		TABLES: [TABLE1, TABLE2] #optional
	env:
		ENABLE_DATETIME_FORMATTING: true
		TIME_FORMAT: HH:mm:ss
		DATE_FORMAT: dd-MM-yyyy
		DATE_TIME_FORMAT: "dd-MM-yyyy HH:mm:ss"
	jdbc:
		oracle.jdbc.timezoneAsRegion: false

db2:
	service:
		DRIVER: MYSQL
		USERNAME: USER
		PASSWORD: PASS
		SERVER: IP / FQDN
		DATABASE: MY_DATABASE
		TABLES: [TABLE1, TABLE2] #optional
	env:
		ENABLE_DATETIME_FORMATTING: true
		TIME_FORMAT: HH:mm:ss
		DATE_FORMAT: dd-MM-yyyy
		DATE_TIME_FORMAT: "dd-MM-yyyy HH:mm:ss"
	jdbc:
		oracle.jdbc.timezoneAsRegion: false
```

Then you can run db2rest like that:

`java --config=datasources.conf -jar db2rest.jar`

When the application starts you can query the different data sources like that

`http://localhost:8090/db1/my_table/?`

`http://localhost:8090/db2/my_view/?`

Something like that.

**Comment by kdhrubo** [comment]
@MichalisDBA @thadguidry - challenge with mutli db is loading meta data. for more than one database it can take a while
Hence proposing loading on demand or incrementaly on first encounter? In this case the first call to table can be slow.

**Reply by kdhrubo** [reply]
This will also make the parameter `INCLUDE_SCHEMAS` redundant.

**Reply by MichalisDBA** [reply]
Sounds good as long as you keep the app fast. Will the metadata for the cached table be kept permanent in memory or they will evict? I mean let's say i query Table1 and the first time it takes 1sec. After that the response time will be 20ms. If someone decide to query the endpoint after an hour, will the endpoint respond in 20ms again or it will have to cache the metadata again?

Also if the metadata are not loaded in the first time, how the `$schema` endpoint will work? It will not have a listing of the tables or views that you can query.

**Reply by kdhrubo** [reply]
https://github.com/kdhrubo/db2rest/issues/554

**Comment by kdhrubo** [comment]
@MichalisDBA closing this one as the feature has been implemented @thadguidry
